ruby_programming/git/advanced_git_lesson.md

```diff
@@ -0,0 +1,192 @@
+
+~~~bash
+pick c85c71dd5 Rewrite Introduction
+squash 4ad2c8df6 Add information on using amend
+squash 729db1da4 Begin work on Remotes
+~~~
+
+That's it! When I save and exit the editor, the bottom two commits on that list get squashed into the top commit. 
+
+## Splitting Up a Commit
+
+Before diving into Remotes, we're going to have a look at a handy Git command called `reset`. Let's say one of my commits was `c85c71dd5 Add knight class and Add attack module`. That's a mouthful, and a bit much for one commit as you should have learned in the previous lesson on commits. So what we're going to do is split it up into two smaller commits, also using the interactive `rebase` tool. 
+
+We open up the tool just like last time, change `pick` to `edit` for the commit we're going to split. Now, however, what we're going to do is run `git reset HEAD^`. What this does is it resets the commit to the one right before HEAD. This allows us to add the files individually, commit them individually, and then `git rebase --continue` just like we did last time to finish up our changes. All together it would look something like this:
+
+~~~bash
+$ git reset HEAD^
+$ git add lib/knight.rb
+$ git commit -m 'Add knight class'
+$ git add lib/movement.rb
+$ git commit -m 'Add attack module'
+$ git rebase --continue
+~~~
+
+Let's start by looking a bit closer at what happened here. When you ran `git reset`, you reset the current branch by pointing HEAD at the commit right before it. At the same time, `git reset` also updated the index (the staging area) with the contents of wherever HEAD now pointed. So your staging area was also reset to what it was at the prior commit - which is great - because this allowed you to add the files we wanted one by now.
+
+Now let's say you want to move where HEAD points to, but *don't* want to touch the staging area. If you want to leave the index alone, you can use `git reset --soft`. This would only perform the first part of `git reset` where the HEAD is moved to point somewhere else.
+
+The last part of reset we want to touch upon is `git reset --hard`. What this does is it performs all the steps of `git reset`, moving the HEAD and updating the index, but it *also* updates the working directory. This is important to note because it can be dangerous, and can potentially destroy data. A hard reset overwrites the files in the working directory to make it look exactly like the staging area of wherever HEAD ends up pointing to. Similarly to `git commit --amend`, a hard reset is a destruction command which overwrites history. This doesn't mean you should completely avoid it if working with shared repositories on a team with other developers. You should, however, **make sure you know exactly why you're using it, and that your coworkers are also aware of how and why you're using it.**
```

Comment:
```suggestion
The last part of reset we want to touch upon is `git reset --hard`. What this does is it performs all the steps of `git reset`, moving the HEAD and updating the index, but it *also* updates the working directory. This is important to note because it can be dangerous, and can potentially destroy data. A hard reset overwrites the files in the working directory to make it look exactly like the staging area of wherever HEAD ends up pointing to. Similarly to `git commit --amend`, a hard reset is a destructive command which overwrites history. This doesn't mean you should completely avoid it if working with shared repositories on a team with other developers. You should, however, **make sure you know exactly why you're using it, and that your coworkers are also aware of how and why you're using it.**
```